### DISCUSSION TITLE
How to use HeightFieldShape.SetHeights

### DISCUSSION THREAD

**Discussion by kutu** [jrouwe/JoltPhysics.js#100] [Q&A]
I pre-generate 9 bodies of HeightFieldShape, to form a square around my character.
When character is moving, I want to re-generate heights for 3 of those bodies and move them in front of my character, to create an infinite world.

I create another one shape, and I reuse its `ArrayFloat` to fill it with new values, to then pass its `FloatMemRef` into `SetHeights` 
(because I don't know how to just create `ArrayFloat` in Jolt, `new Jolt.ArrayFloat()` doesn't work)

```js
shape = new Jolt.HeightFieldShapeSettings();
this.heightSamples = shape.mHeightSamples;
this.heightSamples.resize(totalSize);
```
and then when I need to re-generate one of 9 bodies:
```js
shape = Jolt.castObject(body.GetShape(), Jolt.HeightFieldShape);
sampleCount = shape.GetSampleCount();
totalSize = sampleCount ** 2;
samples = new Float32Array(Jolt.HEAPF32.buffer, Jolt.getPointer(this.heightSamples.data()), totalSize);
samples.set(new Float32Array(data.positionBuffer));
shape.SetHeights(0, 0, sampleCount, sampleCount, this.heightSamples.data(), sampleCount, jolt.GetTempAllocator());
```

In sample `DeformedHeightFieldShapeTest` there are 2 methods called after `SetHeights`, but not available in js

```js
mHeightField->SetHeights(start_x, start_y, count_x, count_y, mHeightSamples.data() + start_y * cSampleCount + start_x, cSampleCount, *mTempAllocator);

// Notify the shape that it has changed its bounding box
mBodyInterface->NotifyShapeChanged(mHeightFieldID, old_com, false, EActivation::DontActivate);

// Activate bodies in the affected area (a change in the height field doesn't wake up bodies)
affected_area.ExpandBy(Vec3::sReplicate(0.1f));
DefaultBroadPhaseLayerFilter broadphase_layer_filter = mPhysicsSystem->GetDefaultBroadPhaseLayerFilter(Layers::MOVING);
DefaultObjectLayerFilter object_layer_filter = mPhysicsSystem->GetDefaultLayerFilter(Layers::MOVING);
mBodyInterface->ActivateBodiesInAABox(affected_area, broadphase_layer_filter, object_layer_filter);
```

**Comment by LeXXik** [comment]
A bit off-topic, but just a note on your project. Jolt (JS) is 32-bit based, so after a few kilometers away from world origin you will start having collision and rendering issues. And even though you could work around the rendering issues, the collision ones will be hard to mitigate. You could probably resort to some tricks, like teleporting the world origin every now and then, but depending on the world complexity it might not be easy.

**Comment by jrouwe** [comment]
Hello,

The SetHeights function should be available, it's here:

https://github.com/jrouwe/JoltPhysics.js/blob/32094603e19a5069eb70c458a96c8ed6c1c9c4e7/JoltJS.idl#L870-L871

You may need to do:

```
Jolt.castObject(shape, Jolt.HeightFieldShape)
```

And w.r.t. Jolt being 32-bit based. I have been thinking to enable JPH_DOUBLE_PRECISION for the library, which switches all positions to double. Downside is that everyone will need to update their JS and change a lot of Vec3 -> RVec3 and Mat44 -> RMat44.

**Reply by LeXXik** [reply]
Hmm. I don't know how much demand there is for a 64-bit version. However, if the demand gets high, perhaps one option could be to have a build variant. But if it cannot be automated, not sure if it is feasible. After all, we mostly operate in 32-bit space when we deal with WebGL.

**Reply by jrouwe** [reply]
I've added the option to compile the library in double precision mode now, by default it isn't though.

**Comment by kutu** [comment]
https://jsfiddle.net/8b39xoyw/2/

Shape created in usual way, vertical bounds [-1, 1]

Then using `SetHeights` bounds stays [-1, 1], but expected to be [-2, 2], even so my temporary `heightSamples` is modified if I log `heightSamples.at(0)`

**Reply by jrouwe** [reply]
If you want to change the height range after creation you need to set the min/max height at construction time:

https://github.com/jrouwe/JoltPhysics.js/blob/32094603e19a5069eb70c458a96c8ed6c1c9c4e7/JoltJS.idl#L853-L854

Note that if you pick a high range, there will be compression artifacts.

In general, I'm not sure if updating the heights of an existing height field is the way to go for this as it is very likely that you'll get compression artifacts. The SetHeights function is meant for editing small parts of the terrain in a limited way. You may be better off just creating a new height field.

**Comment by PhoenixIllusion** [comment]
I saw that the ArrayFloat is missing the constructor.

With the update on Emscripten version, we have the WebID Malloc in the public build, so I tested with that.
https://github.com/jrouwe/JoltPhysics.js/blob/main/Examples/heightfield.html

Clicking on the SVG in the upper SVG of the demo will load up an 8x8 grid of heightmap, lower it by 0.1 float, then load it back to the same spot and also update the backing canvas of the Three heightmap. 

This would go below the [sphere] part, inside the initJolt callback.

```javascript
const READ_SIZE = 8; // must be multiple of block-size used in construction
const terrainShape = Jolt.castObject(terrain.GetShape(), Jolt.HeightFieldShape);
const tempBufferPtr = Jolt._webidl_malloc(7*7*4);
const tempBuffer = new Float32Array(Jolt.HEAPF32.buffer, tempBufferPtr, READ_SIZE*READ_SIZE);// READ_SIZE grid

const canvas = displacementMap.source.data;
const ctx = canvas.getContext('2d');

const svgEle = document.getElementById("sampleImg");
svgEle.addEventListener('click', (mouseEvt) => {
	const x = mouseEvt.offsetX / svgEle.clientWidth;
	const y = mouseEvt.offsetY / svgEle.clientHeight;

	const offsetX = Math.floor( x * 256) - 3; //offset left of click
	const offsetY = Math.floor( y * 256) - 3; //offset up from click
	const texData = ctx.getImageData(offsetX, offsetY, READ_SIZE, READ_SIZE);

	terrainShape.GetHeights(offsetX, offsetY, READ_SIZE, READ_SIZE, tempBufferPtr, READ_SIZE)

	tempBuffer.forEach((_o, i) => {
		tempBuffer[i] -= 0.1;
		for(let j=0;j<3;j++) {
			texData.data[i*4+j] = tempBuffer[i] * 10;
		}
	});

	terrainShape.SetHeights(offsetX, offsetY, READ_SIZE, READ_SIZE, tempBufferPtr, READ_SIZE, jolt.GetTempAllocator());
	ctx.putImageData(texData, offsetX, offsetY);
	displacementMap.needsUpdate = true;
	dynamicObjects.forEach(obj => bodyInterface.ActivateBody(obj.userData.body.GetID())); // wake sleeping objects
})
```

**Reply by jrouwe** [reply]
There's a small mistake where offsetX and Y are not clamped properly and not aligned to the block size. I fixed them and integrated the change into the example. Thanks!

**Reply by PhoenixIllusion** [reply]
Sorry, I noticed the main question was in reference to activating the nearby items, not with regard to setting the height itself.

I saw that the body interface call wasn't in the IDL, so to replicate that likely would need go do a Broad phase AABox query for the nearby region and loop through to wake those up.

The wakes up part of the equation doesn't seem to apply to the original question since it only impacts items currently resting against an actively deforming height field.

If you are going to teleport the height map somewhere else, then you will logically only populate the area with new items after the terrain is teleported (or they will fall into the void below), and every new populated item can be woken up at that time.

**Reply by PhoenixIllusion** [reply]
I'd thought about making it an actual modification to the current example instead of just snippet post, but was going to do a bit more for that. Probably ray-pick to the terrain and modify it that way so we can use a mouse/touch to raise/lower. Probably force up the balls to above any raised edges.

It may need the BodyInterface notify-shape-changed call to properly handle some logic 100% as the Sample, since we can't work around that ourselves, ex: BodyManager clear contact cache, broadphase AABB change if you modify the height too much. I looked into faking it, but calling setShape with itself triggers a pointer check,sees they are the same, and NOP)

**Reply by jrouwe** [reply]
Ok feel free to change it, I just added the NotifyShapeChanged call.
